### PR TITLE
Use ArrayPool in CopyToAsync on several System.IO streams

### PR DESCRIPTION
**Pull request by stephentoub**
- FileStream (already in async on Windows as of my previous PR, now for all flavors)
- DeflateStream
- GZipStream
- PipeStream
- UnmanagedMemoryStream

cc: @JeremyKuhne, @ianhays, @weshaggard, @jkotas, @davidfowl 
https://github.com/dotnet/corefx/issues/11667

### CODE REVIEW

File: src/Common/src/System/IO/StreamHelpers.cs [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+        /// <summary>Standard read/write loop using ReadAsync on the source and WriteAsync on the destination.</summary>
+        private static async Task ArrayPoolCopyToAsyncCore(Stream source, Stream destination, int bufferSize, CancellationToken cancellationToken)
+        {
+            byte[] buffer = ArrayPool<byte>.Shared.Rent(bufferSize);
+            try
+            {
+                while (true)
+                {
+                    int bytesRead = await source.ReadAsync(buffer, 0, buffer.Length, cancellationToken).ConfigureAwait(false);
+                    if (bytesRead == 0) break;
```
Comment (ianhays) [inline src/Common/src/System/IO/StreamHelpers.cs]: Do single line if/bodies like this adhere to the style guide?

**Comment by stephentoub** [comment]
Test Innerloop CentOS7.1 Debug Build and Test please (seg fault in CLI)
@adityamandaleeka, do we think this seg fault is the same one you were tracking?  I've seen it happening on CentOS on PRs several times over the last few days.

**Comment by adityamandaleeka** [comment]
@stephentoub I just hopped on the machine where the failure occurred to take a look. I see that a corerun proceess (46188) crashed with a segfault in JIT code around 61 minutes ago. If that's roughly when this crash occurred, it's a pretty safe bet that it's the mprotect/page migration race I mentioned before.

However, there was also a core dump stored in /tmp/coredumps which seems to be from a prior crash (of PID 11345) that occurred over 2 hours ago. Looking at that dump, the si_code doesn't appear consistent with the symptoms of the kernel bug (which leads to ACCERR).

**Comment by ianhays** [comment]
Looks good!

**Comment by stephentoub** [comment]
Test Innerloop CentOS7.1 Debug Build and Test please (more seg faults on CentOS, unrelated to this PR)
Test Innerloop CentOS7.1 Release Build and Test please (more seg faults on CentOS, unrelated to this PR)

File: src/Common/src/System/IO/StreamHelpers.cs (hunk 1 above)
Comment (stephentoub) [inline src/Common/src/System/IO/StreamHelpers.cs]: I believe so, but I'll make it multi-line, regardless.

**Comment by stephentoub** [comment]
Thanks, @adityamandaleeka.

**Comment by stephentoub** [comment]
Thanks, Ian.

**Pull request merged**